### ISSUE TITLE
feat: addition of generic in `FormDataConsumer`

### ISSUE THREAD

**Issue by Gabriel-Malenowitch** [marmelab/react-admin#8353]
**Is your feature request related to a problem? Please describe.**
 - A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]
 - Currently It turns the code more verbose. it's very common for me to write forms with this component and its became a boring work, because typescript don't get me the suggestions

**Describe the solution you'd like**
 - Adding the generic in the `FormDataConsumer` component
 How it is currently:
```typescript
const FormDataConsumer = (props: ConnectedProps) => {
    const formData = useWatch();

    return <FormDataConsumerView formData={formData} {...props} />;
};
```

How could be to turn our life happier :green_heart: :
```typescript
const FormDataConsumer = <FieldValues extends Record<string, unknown> = {}>(props: ConnectedProps) => {
    const formData = useWatch<FieldValues>();

    return <FormDataConsumerView formData={formData} {...props} />;
};
```

**Describe alternatives you've considered**
 - There are a lot of ways to type something with react and typescript, think big or think small, i just want to use typescript with this component : )
**Additional context**
 - I don't know how I could be more specific, but please let me know if you have any questions.

**Comment by fzaninotto** [comment]
Nice idea for an enhancement! Feel free to open a PR to implement it.

**Comment by Gabriel-Malenowitch** [comment]
@fzaninotto my PR is ready for review. Can you review it pls? https://github.com/marmelab/react-admin/pull/8389

**Comment by viniciusbsneto** [comment]
Looking forward to this PR!

**Comment by lukascivil** [comment]
@fzaninotto help us here, pls!

**Comment by fzaninotto** [comment]
Fixed by #8389
